components/tidb_query/src/codec/convert.rs

```diff
@@ -228,44 +232,65 @@ impl ToInt for u64 {
     fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
         let upper_bound = integer_unsigned_upper_bound(tp);
         if *self > upper_bound {
-            ctx.handle_overflow_err(overflow!(self, upper_bound))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(upper_bound);
         }
         Ok(*self)
     }
 }
 
 impl ToInt for f64 {
+    /// This function port from TiDB's types.ConvertFloatToInt,
+    /// it will check whether the number overflow signed lower and upper bound of `tp`.
+    ///
+    /// # Notes
+    ///
+    /// If overflow, it will handle overflow using ctx,
+    /// so if you get a err returned, needn't to handle it one more time
     fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
+        #![allow(clippy::float_cmp)]
         let val = (*self).round();
         let lower_bound = integer_signed_lower_bound(tp);
         if val < lower_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, lower_bound))?;
+            ctx.handle_overflow_err(overflow(val, tp))?;
             return Ok(lower_bound);
         }
 
         let upper_bound = integer_signed_upper_bound(tp);
-        if val > upper_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, upper_bound))?;
-            return Ok(upper_bound);
+        let ub_f64 = upper_bound as f64;
+        // according to https://github.com/pingcap/tidb/pull/5247
+        if val >= ub_f64 {
+            if val == ub_f64 {
+                return Ok(upper_bound);
+            } else {
+                ctx.handle_overflow_err(overflow(val, tp))?;
+                return Ok(upper_bound);
+            }
         }
         Ok(val as i64)
     }
 
+    /// This function port from TiDB's types.ConvertFloatToUint,
```

Comment:
```
/// This function is ported from TiDB xxxx, which checks whether the number overflows the signed upper and lower boundaries of `tp`
///
/// # Notes
/// 
/// It handles overflows using `ctx` so that the caller would not handle it anymore.
```